subprojects/android-test/ui-testing-core/src/main/kotlin/com/avito/android/test/matcher/OverlapMatcher.kt

```diff
@@ -0,0 +1,52 @@
+package com.avito.android.test.matcher
+
+import android.graphics.Rect
+import android.view.View
+import android.view.ViewGroup
+import org.hamcrest.Description
+import org.hamcrest.TypeSafeMatcher
+
+/**
+ * Check if there is an intersection of the target view's Rect and Rect of any view with higher z-order.
+ *
+ * How it works:
+ * 1. It checks siblings of target view that are below in child list of the parent
+ * 1. Check parent siblings that are below in child list of parent's parent
+ * 1. Check parent's parent siblings... and so on recursively
+ */
+class OverlapMatcher : TypeSafeMatcher<View>() {
+
+    override fun describeTo(description: Description) {
+        description.appendText("overlapped by another views on the same screen")
+    }
+
+    override fun matchesSafely(view: View): Boolean {
+        val viewRect = Rect()
+        view.getGlobalVisibleRect(viewRect)
+
+        return view.parentViewGroup?.iterateViewsBelow(view) { viewBelow ->
+            viewBelow.isIntersectWith(viewRect)
+        } ?: false
+    }
+
+    private fun ViewGroup.iterateViewsBelow(view: View, block: (View) -> Boolean): Boolean {
+        val index = indexOfChild(view)
+        for (i in (index + 1) until childCount) {
+            val child = getChildAt(i)
+            if (block(child)) {
+                return true
+            }
+        }
+
+        return parentViewGroup?.iterateViewsBelow(this, block) ?: false
```

Comment:
I saw an another strategy without recursion. I consider writing code like in that assertion.androidx.test.espresso.assertion.LayoutAssertions#noOverlaps(org.hamcrest.Matcher<android.view.View>)
- You must check on VISIBILITY
I've made a utility function to get all above visible views `com.avito.android.test.util.ViewKt#getAboveVisibleViews`
PR: where I tried to find above clickable view https://github.com/avito-tech/avito-android/pull/146
matcher with simpler hierarchy traverse code https://github.com/avito-tech/avito-android/blob/a240ff41d9498045769abd3f8f5a9229620ff1c1/subprojects/android-test/ui-testing-core/src/main/kotlin/com/avito/android/test/matcher/ViewCoordinatesNotOverlappedMatcher.kt

Reply:
- `com.avito.android.test.util.ViewKt#getAboveVisibleViews` works as expected but return not that views that I need to check for overlapping
- Visibility fixed in 746d623a. Thanks for pointing an issue.